docs(node.js versioning): rewrite (#22714)
renovatebot/renovate: lib/modules/versioning/node/readme.md

Renovate's Node.js versioning is a wrapper around npm versioning.

> But Renovate removes any `v` prefixes from semantic versions when replacing.

Reviewer: ```suggestion
But Renovate removes any `v` prefixes from semantic versions when replacing.

Its primary purpose is to add Node.js LTS awareness, e.g.
- Odd releases are unstable
- Even releases do not reach stability (LTS) immediately
```

Renovate's Node.js versioning is a wrapper around npm's versioning.

> But Renovate removes any `v` prefixes from exact versions when replacing.

Author: It's not clear to me what we mean with `exact` versions here:

- Exact version: `v1.0.0`?
- Loose version: `v1`?
- Exact is the same as pinned: `v1.0.0`?
- Something else

Also _when replacing_, is vague to me as well.

Reviewer 2: I checked all of the scenarios and it didn't matter what the current version was. The `v` prefix is stripped regardless.

Also "replacing" here appears to refer to the `RangeStrategy` used:

https://github.com/renovatebot/renovate/blob/f1f5f270f08711e71e370118ee0e3e642af9a34d/lib/types/versioning.ts#L6

I think it's an abstraction leak from code to documentation. But again, the actual `RangeStrategy` doesn't appear to impact whether or not we remove the `v` prefix.

Author: ### Response to comment

> I checked all of the scenarios and it didn't matter what the current version was. The `v` prefix is stripped regardless.

Okay! So we can just say that Renovate always removes the `v` prefix.

> I think it's an abstraction leak from code to documentation.

I don't understand what you're saying here. I think you mean that "translating code to human docs text" went wrong somewhere?

> But again, the actual `RangeStrategy` doesn't appear to impact whether or not we remove the `v` prefix.

Thanks for explaining that. I got confused because the `rangeStrategy` docs mention `pin` and `replace` behaviors. 😄 

### Suggestion to review/edit

```suggestion
If the version has a `v` prefix, Renovate always removes it when creating the update.
```

We plan to extend the Node.js versioning to support "stability" awareness.

> This is nice to have because Node.js's version stability does not match the SemVer rules.

Author: Is there any issue we can link to here for the plan to extend Node.js versioning to support "stability" awareness? I searched a bit, but could not find a 1 to 1 match. 😄

Reviewer 2: I couldn't find an issue for Node.js stability specifically. The closest I think matches is https://github.com/renovatebot/renovate/issues/5990

Author: I think I found the source commit for the `node/readme.md` file. Here's a quote from that commit: [^1]

> ## Implementation plan/status
> 
> - [x] Strip v prefix
> - [ ] Support Node.js-specific "stable" awareness

Multiple files in that commit have a section like this. So I think the whole "We plan to extend ..." stuff was a developer todo-list, instead of something that should go in the end-user documentation.

> I couldn't find an issue for Node.js stability specifically. The closest I think matches is #5990

I don't know if we need a new issue just for "Node.js stability awareness" or if we should link to #5990 in the docs. I'll let @rarkins decide what to do here. 😉 

[^1]: https://github.com/renovatebot/renovate/commit/e5cd1cd4b0beace2d1f88280ca2bc66d9ce2a752#diff-ecd41b3a4fd2acb98436c3d49e330ec141e1b18875c5b8bd618d3428546c9df8

Author: ```suggestion
```

Renovate's Node.js versioning is a wrapper around npm versioning.
But Renovate removes any `v` prefixes from semantic versions when replacing.

Its primary purpose is to add Node.js LTS awareness, e.g.:

- Odd releases are unstable

> - Even releases do not reach stability (LTS) immediately

Reviewer: ```suggestion
- Even releases do not reach stability (LTS) immediately

Unfortunately there is not currently any way to get `node` versioning to replace `docker` versioning when Docker image tags contain a compatibility suffix such as `-alpine` or `-slim`, because npm versioning treats these suffixes as implying pre-releases.
```

Renovate's Node.js versioning is a wrapper around npm versioning.
But Renovate removes any `v` prefixes from semantic versions when replacing.

Its primary purpose is to add Node.js LTS awareness, e.g.:

- Odd releases are unstable
- Even releases do not reach stability (LTS) immediately

> Unfortunately, there is not currently any way to get `node` versioning to replace `docker` versioning when Docker image tags contain a compatibility suffix such as `-alpine` or `-slim`, because npm versioning treats these suffixes as implying pre-releases.

Author: This sentence is too long, and has too many conditions in it.

Can we chop things up like this, maybe? We should also say upfront _who_ is the targe audience for the paragraph.

```suggestion
If you use Docker, read this paragraph.
You can _not_ use `node` versioning to replace `docker` versioning.
This is because npm versioning treats these suffixes as implying pre-releases.
Using `node` versioning would helped you when Docker image tags have a compatibility suffix like `-alpine` or `-slim`.
```
